server/src/main/java/org/opensearch/common/time/JavaDateFormatter.java

```diff
@@ -285,27 +286,42 @@ public TemporalAccessor parse(String input) {
      * @throws DateTimeParseException when unable to parse with any parsers
      */
     private TemporalAccessor doParse(String input) {
-        if (parsers.size() > 1) {
-            Object object = null;
-            if (canCacheLastParsedFormatter && lastParsedformatter != null) {
-                ParsePosition pos = new ParsePosition(0);
-                object = lastParsedformatter.toFormat().parseObject(input, pos);
+        if (parsers.size() != 1) {
+            // Copy to local from volatile to avoid error in multi-threads,
+            // other threads may update lastParsedformatter
+            final DateTimeFormatter cachedFormatter = lastParsedformatter;
+            if (canCacheLastParsedFormatter && cachedFormatter != null) {
+                final ParsePosition pos = new ParsePosition(0);
+                Object object = tryParseObject(cachedFormatter, input, pos);
                 if (parsingSucceeded(object, input, pos)) {
                     return (TemporalAccessor) object;
                 }
             }
             for (DateTimeFormatter formatter : parsers) {
-                ParsePosition pos = new ParsePosition(0);
-                object = formatter.toFormat().parseObject(input, pos);
-                if (parsingSucceeded(object, input, pos)) {
-                    lastParsedformatter = formatter;
-                    return (TemporalAccessor) object;
+                if (formatter != cachedFormatter || !canCacheLastParsedFormatter) {
+                    final ParsePosition pos = new ParsePosition(0);
+                    Object object = tryParseObject(formatter, input, pos);
+                    if (parsingSucceeded(object, input, pos)) {
+                        if (canCacheLastParsedFormatter) {
+                            lastParsedformatter = formatter;
+                        }
+                        return (TemporalAccessor) object;
+                    }
                 }
             }
 
-            throw new DateTimeParseException("Failed to parse with all enclosed parsers", input, 0);
+            throw new DateTimeParseException("failed to parse date field [" + input + "] with format [" + format + "]", input, 0);
+        } else {
+            return parsers.get(0).parse(input);
+        }
+    }
+
+    private Object tryParseObject(final DateTimeFormatter formatter, final String input, final ParsePosition pos) {
```

Comment:
Could you please add javadocs referencing https://bugs.openjdk.org/browse/JDK-8319640 so others would know why we are doing this, thank you